### PR TITLE
Add caching

### PR DESCRIPTION
**Pull request by bclymer**
I took a lot of the logic from the previous parallel PR actually. Specifically the fact that `processGame` returns `data` and then you `processResults` separately. We store only the data we return in `data` in the JSON file, as well as the current app version from `package.json `and `slippiJs` version from `package-lock.json`.

I just take the MD5 of the file name for the hash. If somebody renames a different replay to look like another replay, the deserve to have weird data.

### CODE REVIEW

File: slippi-stats.js [hunk 1]
```diff
@@ -1,14 +1,22 @@
...
 
 // Characters ordered by ID
-var characters = ['Captain Falcon', 'Donkey Kong', 'Fox', 'Mr. Game & Watch', 'Kirby', 'Bowser',
+const  characters = ['Captain Falcon', 'Donkey Kong', 'Fox', 'Mr. Game & Watch', 'Kirby', 'Bowser',
             'Link', 'Luigi', 'Mario', 'Marth', 'Mewtwo', 'Ness', 'Peach', 'Pikachu',
             'Ice Climbers', 'Jigglypuff', 'Samus', 'Yoshi', 'Zelda', 'Sheik', 'Falco',
             'Young Link', 'Dr. Mario', 'Roy', 'Pichu', 'Ganondorf']
 
-var characters_lowercase = ['captain falcon', 'donkey kong', 'fox', 'mr. game & watch', 'kirby', 'bowser',
+            const  characters_lowercase = ['captain falcon', 'donkey kong', 'fox', 'mr. game & watch', 'kirby', 'bowser',
```
Comment (brandoncollins92) [inline slippi-stats.js:19]:
```suggestion
const characters_lowercase = ['captain falcon', 'donkey kong', 'fox', 'mr. game & watch', 'kirby', 'bowser',
```

File: slippi-stats.js [hunk 2]
```diff
@@ -1,14 +1,22 @@
...
+const jsonLock = require('./package-lock.json')
+const crypto = require('crypto')
+
+const statsVersion = pjson.version
+const slippiJsVersion = jsonLock.dependencies["@slippi/slippi-js"].version
+const cacheFilePath = "./cacheFile.json"
 
 // Characters ordered by ID
-var characters = ['Captain Falcon', 'Donkey Kong', 'Fox', 'Mr. Game & Watch', 'Kirby', 'Bowser',
+const  characters = ['Captain Falcon', 'Donkey Kong', 'Fox', 'Mr. Game & Watch', 'Kirby', 'Bowser',
```
Comment (brandoncollins92) [inline slippi-stats.js:14]:
```suggestion
const characters = ['Captain Falcon', 'Donkey Kong', 'Fox', 'Mr. Game & Watch', 'Kirby', 'Bowser',
```

File: slippi-stats.js [hunk 3]
```diff
@@ -42,6 +50,19 @@ if (character_arg) {
...
+function loadCache() {
+    try {
+        const contents = fs.readFileSync(cacheFilePath, 'utf8')
+        const data = JSON.parse(contents)
+        if (!data) { return }
+        if (data.statsVersion != statsVersion) { return }
+        if (data.slippiJsVersion != slippiJsVersion) { return }
+        return data.results
+    } catch {
+        return undefined
```
Comment (brandoncollins92) [inline slippi-stats.js:62]: Any reason you're returning `undefined` here, but doing an empty return a few lines above? Functionally they're the same, but for consistency's sake they might as well be the same ?

File: slippi-stats.js [hunk 4]
```diff
@@ -91,33 +112,56 @@ var final_player_name = user_player
```
Comment (BrodyVoth) [inline slippi-stats.js:112]:
```suggestion
let real_player_code
let final_opponent_name
let real_opponent_code
```
Some searches fail cause of uninitialized variables at the start like real_player_code, final_opponent_name, real_opponent_code (like you fixed in 3fd6e39)

File: slippi-stats.js [hunk 5]
```diff
@@ -217,157 +261,196 @@ for (i = 0; i < files.length; i++) {
...
+        data.stage_num = settings.stageId
+
         // Try to find last used nickname and actual connect code to display at the end
         if (player_name.length > 0) {
-            final_player_name = player_name
+            data.final_player_name = player_name
         }
-        real_player_code = player_codes[player_num]
+        data.real_player_code = player_codes[player_num]
         if (opponent_arg && player_names[opponent_num]) {
```
Comment (BrodyVoth) [inline slippi-stats.js:281]:
```suggestion
        if (player_names[opponent_num]) {
```
Remove opponent_arg check before saving final_opponent_name or opponents' names will only be saved if doing a head to head search

**Comment by BrodyVoth** [comment]
The new implementation is great. I'm going to merge and tweak a couple things, like not trying to slowly save stats for games that have no player data.

**Pull request merged**